src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -0,0 +1,530 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * Copyright (c) 2019 ifak e.V. Magdeburg (Holger Zipper)
+ * Copyright (c) 2022 Linutronix GmbH (Author: Muddasir Shakil)
+ */
+
+#include "ua_pubsub_keystorage.h"
+
+#if defined(UA_ENABLE_PUBSUB_ENCRYPTION) && defined(UA_ENABLE_PUBSUB_SKS) /* conditional compilation */
+
+#define UA_CURRENTKEY_COUNT 1
+
+#include "server/ua_server_internal.h"
+
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_new(void) {
+    UA_PubSubKeyStorage *keyStorage;
+
+    /*allocate memory for Keystorage*/
+    keyStorage = (UA_PubSubKeyStorage *)UA_malloc(sizeof(UA_PubSubKeyStorage));
+    if(!keyStorage)
+        return NULL;
+
+    /*initialize it to have the possibility to delete it in case of error somewhere*/
+    memset(keyStorage, 0, sizeof(UA_PubSubKeyStorage));
+
+    return keyStorage;
+}
+
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_getKeyStorageWithSecurityGroupId(UA_Server *server,
+                                                     UA_String securityGroupId) {
+
+    if(!server || UA_String_isEmpty(&securityGroupId))
+        return NULL;
+
+    UA_PubSubKeyStorage *outKeyStorage, *keyStorageTemp;
+    LIST_FOREACH_SAFE(outKeyStorage, &server->pubSubManager.pubSubKeyList, keyStorageList,
+                      keyStorageTemp) {
+        if(UA_String_equal(&outKeyStorage->securityGroupID, &securityGroupId)) {
+            return outKeyStorage;
+        }
+    }
+    return NULL;
+}
+
+UA_StatusCode
+UA_PubSubKeyStorage_getPubSubSecurityPolicyByUri(UA_Server *server,
+                                                 const UA_String *securityPolicyUri,
+                                                 UA_PubSubSecurityPolicy **policy) {
+    if(!server || !securityPolicyUri)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    UA_StatusCode retval = UA_STATUSCODE_BAD;
+
+    UA_ServerConfig *config = UA_Server_getConfig(server);
+
+    for(size_t i = 0; i < config->pubSubConfig.securityPoliciesSize; i++) {
+        if(UA_String_equal(securityPolicyUri,
+                           &config->pubSubConfig.securityPolicies[i].policyUri)) {
+            *policy = &config->pubSubConfig.securityPolicies[i];
+            retval = UA_STATUSCODE_GOOD;
+            break;
+        }
+    }
+    return retval;
+}
+
+static void
+UA_PubSubKeyStorage_clearKeyList(UA_PubSubKeyStorage *keyStorage) {
+    if(keyStorage->keyList) {
+        UA_PubSubKeyListItem *iterator = keyStorage->firstItem;
+        UA_PubSubKeyListItem *head = iterator;
+        while(iterator != NULL) {
+            if(iterator->key.data) {
+                UA_free(iterator->key.data);
```

Comment:
UA_ByteString_clear.
That covers the difference between "empty string" and "null string" in the SDK.